### PR TITLE
Add support for touch-based graph navigation and selection

### LINKED ISSUE #9493: Pinch to zoom doesn't work on windows touchscreen
My windows laptop has a touch screen which allows pinch to zoom in other applications, but it doesn't work in the Enso graph editor.

### PR DESCRIPTION
**Pull request by Frizi**
### Pull Request Description

Fixes #9493

Tested on windows desktop (standard mouse), macbook touchpad and iphone, which should behave similarily to windows touchscreen devices.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [x] Unit tests have been written where possible.

### CODE REVIEW

File: app/gui2/src/components/GraphEditor/GraphNode.vue [hunk 1]
```diff
@@ -262,10 +263,15 @@ const dragPointer = usePointer(
         startEpochMs.value = Number(new Date())
         significantMove.value = false
         break
-      case 'stop': {
+      case 'stop':
         startEpochMs.value = 0
         emit('draggingCommited')
-      }
+        break
+
```
Comment (farmaazon) [inline app/gui2/src/components/GraphEditor/GraphNode.vue]: There is no newline before previous case. Let's unify somehow.

File: app/gui2/src/composables/navigator.ts [hunk 2]
```diff
@@ -236,55 +350,18 @@ export function useNavigator(
     () => `translate(${translate.value.x * scale.value}px, ${translate.value.y * scale.value}px)`,
   )
 
-  let isPointerDown = false
-  let scrolledThisFrame = false
+  // let isPointerDown = false
```
Comment (farmaazon) [inline app/gui2/src/composables/navigator.ts]: Commented code.

File: app/gui2/src/composables/navigator.ts [hunk 3]
```diff
@@ -350,45 +427,25 @@ export function useNavigator(
...
-    },
-    pointerEventsCapture: wheelEventsCapture,
     keyboardEvents: panArrows.events,
     translate: readonly(translate),
     targetCenter: readonly(targetCenter),
     targetScale: readonly(targetScale),
     scale: readonly(toRef(scale, 'value')),
     viewBox: readonly(viewBox),
     transform: readonly(transform),
+    addHoldDragListener(listener: Handler<'drag'>) {
```
Comment (farmaazon) [inline app/gui2/src/composables/navigator.ts:444]: A little comment would be useful saying, what exactly is "hold drag" (i.e. it is drag after holding, not somewhat "holding the drag")

File: app/gui2/src/composables/navigator.ts [hunk 4]
```diff
@@ -54,18 +65,143 @@ export function useNavigator(
...
+          scrollTo(center.value.addScaled(Vec2.FromTuple(state.delta), -1 / scale.value))
+          state.event.stopImmediatePropagation()
+        } else if (!isTouch && (state.buttons & PointerButtonMask.Secondary) != 0) {
+          const prevScale = scale.value
+          updateScale((oldValue) => oldValue * Math.exp(-state.delta[1] / 100))
+          scrollTo(
+            center.value
+              .sub(gesturePivot)
+              .scale(prevScale / scale.value)
+              .add(gesturePivot),
```
Comment (farmaazon) [inline app/gui2/src/composables/navigator.ts]: Duplicated code with one section in `onPinch`

File: app/gui2/src/composables/navigator.ts [hunk 5]
```diff
@@ -54,18 +65,143 @@ export function useNavigator(
...
+          holdDragStarted ||
+          (!isTouch && (state.buttons & PointerButtonMask.Main) != 0)
+        ) {
+          callHoldDragListeners({
+            ...state,
+            first: holdDragStarted === false,
+            dragging: state.dragging || !state.last,
+          })
+          holdDragStarted = true
+        }
```
Comment (farmaazon) [inline app/gui2/src/composables/navigator.ts]: If possible, I would put each branch into its own named function. Or, if it's not practical, at least the conditions should be assigned to variables named like `isScrolling`, 'isZooming' etc.

File: app/gui2/src/composables/navigator.ts [hunk 6]
```diff
@@ -54,18 +65,143 @@ export function useNavigator(
...
+      },
+      wheel: {
+        enabled: true,
+      },
     },
   )
 
+  watch(size, updateViewportRect, { immediate: true })
+
+  const targetScale = shallowRef(1)
```
Comment (farmaazon) [inline app/gui2/src/composables/navigator.ts]: Why we need to make ref shallow, if it's only a number? Should all primitives' refs be explicitly shallow for some reason?

File: app/gui2/src/composables/navigator.ts [hunk 7]
```diff
@@ -54,18 +65,143 @@ export function useNavigator(
...
+              .sub(gesturePivot)
+              .scale(prevScale / scale.value)
+              .add(gesturePivot),
+          )
+        } else if (
+          state.tap ||
+          holdDragStarted ||
+          (!isTouch && (state.buttons & PointerButtonMask.Main) != 0)
+        ) {
+          callHoldDragListeners({
```
Comment (farmaazon) [inline app/gui2/src/composables/navigator.ts]: This "holdDrag" is not always "hold drag" - sometimes it's just mouse-clik movement.

(Oh, well, one could interpret it as both "drag after hold" and "drag with button hold" ;) but then it requires docs somewhere.

**Review by farmaazon — APPROVED** [review]
I think navigator needs some tests, as the logic there looks quite complex, and dragging/pinching won't be tested routinely.

As it's a bit of PR, we can make it in a separate one, or even as a separate task. Otherwise, I see mostly naming issues.

File: app/gui2/src/composables/navigator.ts (hunk 7 above)
Comment (Frizi) [reply app/gui2/src/composables/navigator.ts]: Added docs to the "add hold drag listener" exported function, which exactly points out those conditions.

File: app/gui2/src/composables/navigator.ts (hunk 5 above)
Comment (Frizi) [reply app/gui2/src/composables/navigator.ts]: The condition do depend on the order in the if-else chain. I think refactoring each individual handler case to a separate function makes it very clear what's happening though.

**Review by somebody1234 — APPROVED** [review]
approving dashboard change (removing `@ts-expect-error`)

**Pull request merged**